bundles/org.openhab.binding.echonetlite/src/main/java/org/openhab/binding/echonetlite/internal/EchonetChannel.java

```diff
@@ -0,0 +1,107 @@
+/**
+ * Copyright (c) 2010-2022 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.echonetlite.internal;
+
+import java.io.IOException;
+import java.net.Inet4Address;
+import java.net.InetSocketAddress;
+import java.net.NetworkInterface;
+import java.net.SocketAddress;
+import java.net.StandardProtocolFamily;
+import java.nio.ByteBuffer;
+import java.nio.channels.DatagramChannel;
+import java.nio.channels.SelectionKey;
+import java.nio.channels.Selector;
+import java.util.Enumeration;
+import java.util.function.Consumer;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Wraps a Datagram channel for sending/receiving data to/from echonet lite devices.
+ *
+ * @author Michael Barker - Initial contribution
+ */
+@NonNullByDefault
+public class EchonetChannel {
+
+    private final Logger logger = LoggerFactory.getLogger(EchonetChannel.class);
+
+    private final DatagramChannel channel;
+    private final Selector selector = Selector.open();
+
+    private short tid = 0;
+
+    public EchonetChannel(InetSocketAddress discoveryAddress) throws IOException {
+        channel = DatagramChannel.open(StandardProtocolFamily.INET);
+        channel.bind(new InetSocketAddress("0.0.0.0", discoveryAddress.getPort()));
+        final Enumeration<NetworkInterface> networkInterfaces = NetworkInterface.getNetworkInterfaces();
+        while (networkInterfaces.hasMoreElements()) {
+            final NetworkInterface networkInterface = networkInterfaces.nextElement();
```

Comment:
To fix `[INFO] EchonetClass.java:[22,31] Unsafe null type conversion (type annotations): The value of type 'org.openhab.binding.echonetlite.internal.Epc.@NonNull Device @org.eclipse.jdt.annotation.NonNull[]' is made accessible using the less-annotated type 'org.openhab.binding.echonetlite.internal.Epc @org.eclipse.jdt.annotation.NonNull[]'`:
```suggestion
            final @Nullable NetworkInterface networkInterface = networkInterfaces.nextElement();
```

Reply:
Again this makes the code look weird as we mark the variable as `@Nullable` and use it without a null check (and adding the null check generates the redundant check warning).  The cleanest solution I've come across is to just cast it to the same type, e.g.

```suggestion
            final NetworkInterface networkInterface = (NetworkInterface) networkInterfaces.nextElement();
```

Which due to the `@NonNullByDefault` is effectively `(@NonNull NetworkInterface)` which is enough to get Spotbugs to stop complaining.  This does come across as a redundant cast in Intellij, but I can live with that.  If necessary we can use the `//NoInspection RedundantCast`, but I've left that out for now.  I tried using `@SupressWarnings("RedundantCast")`, but Spotbugs didn't like that much either.

Reply:
Well, not exactly. This was not a compiler warning, just info - caused by missing null annotation of the `NetworkInterface` classes. By adding `@Nullable` in declaration we have compatible annotation on both sides. This doesn't mean we need to check for null, since probably it was analyzed that `nextElement()` cannot return `null` when `hasMoreElement()` returns `true`.

Reply:
I now realize that `nextElement()` will throw `NoSuchElementException` when there are no more elements, so it cannot return null. Thus, my comment is invalid and your approach is solid. :-)